### ISSUE TITLE
Big tagger delay in refreshing squiggles in the editor

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#65502]
~**This is regression in 17.5 dogfood builds, does not repro on 17.4 release builds**~ This seems to repro on both 17.4.1 and 17.5

**Version Used**:

```
Microsoft Visual Studio Enterprise 2022
Version 17.5.0 Preview 2.0 [33116.440.main]
VisualStudio.17.IntPreview/17.5.0-pre.2.0+33116.440.main
Microsoft .NET Framework
Version 4.8.09032
```

**Steps to Reproduce**:

1. Open Roslyn.sln
2. Open source file `Compilation.cs` in `Microsoft.CodeAnalysis.csproj`
3. Goto line 115 and repeatedly remove and add back the semicolon at the end of the statement in a loop

**Expected Behavior**:
Error list and squiggles in the editor should refresh at a reasonable rate

**Actual Behavior**:
Error list refreshes at a reasonable rate, but squiggles in the editor are extremely slow to refresh, almost taking 10-15 seconds after each keystroke.

Small repros of the experience in 17.5 vs 17.4 builds in the below demo gifs. I also have perf traces for both the below experiences (ping me offline if you need them), but they have very similar CPU stacks, indicating it maybe a regression in the platform layer where the delay at which the taggers are invoked is much higher than earlier (someone with more background on tagger would need to confirm though).

### 17.5 Experience (slow)
![17 5_TaggerDelay](https://user-images.githubusercontent.com/10605811/202671854-59f71e0b-c882-4940-94c7-7a75331cca05.gif)

### 17.4 Experience (fast)
![17 4 0_TaggerDelay](https://user-images.githubusercontent.com/10605811/202671941-c6504157-fcae-4a97-9d72-67b6992353db.gif)

**Comment by CyrusNajmabadi** [comment]
Do you have the lsp experience enabled?

**Comment by mavasani** [comment]
> Do you have the lsp experience enabled?

Nope

**Comment by CyrusNajmabadi** [comment]
The only difference in aware of is that we switched from push diags here to pull (even outside of lsp). I wonder if something got much worse there. 

Might also need to check if the diags are going as slow as the slowest link (e.g. semantic diags).  We may want individual taggers for syntactic vs semantic.

**Comment by CyrusNajmabadi** [comment]
@mavasani we're currently using this:

```c#
        Task<ImmutableArray<DiagnosticData>> GetDiagnosticsForSpanAsync(TextDocument document, TextSpan? range, Func<string, bool>? shouldIncludeDiagnostic, bool includeCompilerDiagnostics, bool includeSuppressedDiagnostics = false, CodeActionRequestPriority priority = CodeActionRequestPriority.None, Func<string, IDisposable?>? addOperationScope = null, CancellationToken cancellationToken = default);
```

Is there an API that gets us things like just syntax-diagnostics so that we can spearate out syntax/semantics/analyzer diagnostics?

**Comment by mavasani** [comment]
> @mavasani we're currently using this:
> 
> ```cs
>         Task<ImmutableArray<DiagnosticData>> GetDiagnosticsForSpanAsync(TextDocument document, TextSpan? range, Func<string, bool>? shouldIncludeDiagnostic, bool includeCompilerDiagnostics, bool includeSuppressedDiagnostics = false, CodeActionRequestPriority priority = CodeActionRequestPriority.None, Func<string, IDisposable?>? addOperationScope = null, CancellationToken cancellationToken = default);
> ```
> 
> Is there an API that gets us things like just syntax-diagnostics so that we can spearate out syntax/semantics/analyzer diagnostics?

There is none right now, but definitely seems we would need one. In fact, the push Disha background analysis executes in 4 steps as below, and we might have to do the same:

1. Compiler syntax diagnostics 
2. Analyzer syntax diagnostics
3. Compiler semantic diagnostics
4. Analyzer semantic diagnostics

Do we return streaming diagnostics or would the platform need separate APIs for syntax and semantic diagnostics and probably even separate requests for compiler and analyzer diagnostics?

**Comment by mavasani** [comment]
So it seems this also repros in 17.4.1 for customers: https://developercommunity.visualstudio.com/t/Extremely-slow-to-update-real-time-compi/10204121#T-N10206358-N10206847

we are seeing lot of customer reports for very slow diagnostic refresh in 17.4.1 and if it turns out to be the same issue then we might need to take this for servicing

**Comment by CyrusNajmabadi** [comment]
> Do we return streaming diagnostics or would the platform need separate APIs for syntax and semantic diagnostics and probably even separate requests for compiler and analyzer diagnostics?

Our intent with LSP diagnostics is to effectively have separate requests, and this is supported by the LSP spec and is coming soon from VS LSP client.  However, that is contingent on LSP diagnostics being turned on by default.

The situation here is the non-LSP case where we own the taggers.  We'll likely have to have 3 squiggle taggers here.  Streaming wouldn't wokr as taggers only show updates when the full set of new tags is computed.  So we need to decouple them.

**Comment by mavasani** [comment]
@CyrusNajmabadi one thing to call out is error list is refreshing at expected fast rates. Only the squiggles in the editor are extremely slow to refresh. Is the Error list still based off push diagnostics?

**Comment by CyrusNajmabadi** [comment]
> There is none right now, but definitely seems we would need one. In fact, the push diagnostics background analysis executes in 4 steps as below, 

Good to know.  We can like likely do the same for taggers by just having 4 of them, all subclassing a common type.

**Comment by CyrusNajmabadi** [comment]
> Only the squiggles in the editor are extremely slow to refresh. Is the Error list still based off push diagnostics?

Yes, error list is push based still.  But will go to pull when we turn on LSP.  It will be based off of the N diagnostic requests though, so syntactic stuff should happen first for both squiggle/error list, then the rest, etc.

**Comment by CyrusNajmabadi** [comment]
I'll look into this on monday.  @arkalyanms for possible servicing need, i anticipate any fix here being small in scope and low risk.

**Comment by mavasani** [comment]
> I'll look into this on monday. @arkalyanms for possible servicing need, i anticipate any fix here being small in scope and low risk.

@CyrusNajmabadi we may want to consider just reverting the taggers back to push diagnostics for 17.4 servicing (if possible) and do the correct fix for pull diags in 17.5

**Comment by CyrusNajmabadi** [comment]
Yup.  That's an option.  PR was https://github.com/dotnet/roslyn/pull/63858.  not sure if we've heard neough to warrant a full rollback.

**Comment by mavasani** [comment]
Another customer report of the same issue on 17.4: https://developercommunity.visualstudio.com/t/Error-highlighting-is-frequently-incorre/10204101

**Comment by mavasani** [comment]
Another potential report based on the provided screenshot (waiting for user confirmation): https://developercommunity.visualstudio.com/t/What-is-wrong-with-your-Intellisense/10202223

**Comment by mavasani** [comment]
https://developercommunity.visualstudio.com/t/COMPILER-SUCKS-IN-2022-1740/10196022 is very likely the same issue as well.

**Comment by CyrusNajmabadi** [comment]
Revert PR available here https://github.com/dotnet/roslyn/pull/65538.

@arkalyanms What should we do to work through the servicing process?

**Comment by arkalyanms** [comment]
Will follow up on the revert PR. Thanks!

**Comment by mavasani** [comment]
Another customer report for the same issue: https://developercommunity.visualstudio.com/t/Errors-are-not-cleared-in-code-window/10206510.

@CyrusNajmabadi @akhera99 It seems many customers that report this stale squiggles issue also see #65503. I am wondering if #65503 also manifests due to this issue of switching to LSP pull diagnostics for tagger OR would we need a separate fix + backport for #65503.

**Comment by Youssef1313** [comment]
Also on Twitter: https://twitter.com/mariusk007/status/1594930592545579008

**Comment by mavasani** [comment]
Reported in https://developercommunity.visualstudio.com/t/Intellisense-not-working-consistently-17/10200700#T-ND10209377

**Comment by mavasani** [comment]
Reported in https://developercommunity.visualstudio.com/t/intellisense-not-working-Please-fix-this/10208337

**Comment by mavasani** [comment]
Slow squiggles refresh - https://developercommunity.visualstudio.com/t/Do-not-redline-faulty-objects-and-sugges/10206065

**Comment by mavasani** [comment]
P0 bug reported at: https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1686342

**Comment by reduckted** [comment]
Also reported here: https://developercommunity.visualstudio.com/t/IntelliSense-for-NET-has-got-slow-after/10211080

**Comment by CyrusNajmabadi** [comment]
This is not likely the same issue. Pull diagnostics would not lead to a minute of waiting. We'd need a trace to set what's going on.

**Comment by reduckted** [comment]
It's been closed as a duplicate of this issue.

**Comment by mavasani** [comment]
> It's been closed as a duplicate of this issue.

All the symptoms in the feedback ticket match this one. @reduckted can you please quickly check if the errors are updated at properly and at a fast rate in the error list while you type?

**Comment by CyrusNajmabadi** [comment]
Feedback ticket says:  "I must wait half a minute until IntelliSense displays it."  that's *def* out of the realm of anything i would expect (Even with pull diagnostics).  We def need a trace here to figure out what's gone so wrong :)

**Comment by reduckted** [comment]
> can you please quickly check if the errors are updated properly and at a fast rate in the error list while you type?

I posted a screen recording in [this comment](https://developercommunity.visualstudio.com/t/IntelliSense-for-NET-has-got-slow-after/10211080#T-N10212555) that might help. I'll attach it here as well: 
https://user-images.githubusercontent.com/10321525/204069941-33e613b6-4657-4216-b6f0-cb371aeee820.mp4

In that scenario, all I was doing was adding line breaks. The error list always showed the errors, but the squiggly lines disappeared.

Here's another recording I just made:
https://user-images.githubusercontent.com/10321525/204070127-815aacb9-4d38-45df-a7f5-b15278c750a6.mp4

In this recording I am creating and fixing an error. The error list and squiggles are all updated reasonably fast, right up until around the 1:15 mark where I make an edit and the squiggles just disappear. The error list continues to show the error. It's been over five minutes since I finished that recording and the squiggles have not come back.

I'm happy to provide traces. I'll just need instructions on how to do that. 😃

**Comment by mavasani** [comment]
So seems there are 2 issues here:

1. With 17.4, squiggles are being refreshed after all compiler and analyzer syntax and semantic diagnostics have finished computing. Earlier it would compute compiler syntax errors upfront and refresh it instantly. This is being tracked by this issue, and porting the fix for this issue to 17.4 should restore squiggles refreshing quickly again.
2. Few minutes delay in computing compiler semantic diagnostics and all analyzer diagnostics is still very large, likely some very slow analyzer is causing it. Can you please file a separate ticket with performance trace to help us identify this slow analyzer? Kindly add a hyperlink to this comment in your feedback ticket so it is easy to correlate. Thanks.

**Comment by reduckted** [comment]
> Can you please file a separate ticket with performance trace to help us identify this slow analyzer?

I've just tried doing this through "Report a problem" and as soon as I edit the file to cause a compilation error, an internal error occurs and no errors are shown.

![image](https://user-images.githubusercontent.com/10321525/204074817-62240fe9-169a-4b90-8fee-eb1b003592ff.png)

> Feature ‘Diagnostic analyzer runner’ is currently unavailable due to an internal error.

```
StreamJsonRpc.RemoteInvocationException: Deserializing JSON-RPC argument with name "" and position 1 to type "" failed: Failed to deserialize System.Object value.
   at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__145`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Remote.BrokeredServiceConnection`1.<TryInvokeAsync>d__17`1.MoveNext()
RPC server exception:
StreamJsonRpc.RpcArgumentDeserializationException: Deserializing JSON-RPC argument with name "" and position 1 to type "" failed: Failed to deserialize System.Object value.
 ---> MessagePack.MessagePackSerializationException: Failed to deserialize System.Object value.
 ---> MessagePack.MessagePackSerializationException: Invalid primitive bytes.
      at MessagePack.Formatters.PrimitiveObjectFormatter.Deserialize(MessagePackReader& reader, MessagePackSerializerOptions options)
      at MessagePack.Formatters.PrimitiveObjectFormatter.DeserializeMap(MessagePackReader& reader, Int32 length, MessagePackSerializerOptions options)
      at MessagePack.Formatters.PrimitiveObjectFormatter.Deserialize(MessagePackReader& reader, MessagePackSerializerOptions options)
      at MessagePack.Formatters.PrimitiveObjectFormatter.DeserializeMap(MessagePackReader& reader, Int32 length, MessagePackSerializerOptions options)
      at MessagePack.Formatters.PrimitiveObjectFormatter.Deserialize(MessagePackReader& reader, MessagePackSerializerOptions options)
      at MessagePack.Formatters.PrimitiveObjectFormatter.Deserialize(MessagePackReader& reader, MessagePackSerializerOptions options)
      at MessagePack.MessagePackSerializer.Deserialize[T](MessagePackReader& reader, MessagePackSerializerOptions options)
   --- End of inner exception stack trace ---
      at MessagePack.MessagePackSerializer.Deserialize[T](MessagePackReader& reader, MessagePackSerializerOptions options)
      at StreamJsonRpc.MessagePackFormatter.JsonRpcRequest.TryGetArgumentByNameOrIndex(String name, Int32 position, Type typeHint, Object& value)
   --- End of inner exception stack trace ---
      at StreamJsonRpc.MessagePackFormatter.JsonRpcRequest.TryGetArgumentByNameOrIndex(String name, Int32 position, Type typeHint, Object& value)
      at StreamJsonRpc.JsonRpcEventSource.GetArgumentsString(JsonRpcRequest request)
      at StreamJsonRpc.JsonRpc.DispatchIncomingRequestAsync(JsonRpcRequest request)
```

It seems I'm not the first to encounter this: https://developercommunity.visualstudio.com/t/Intellisense-in-VS2022-takes-60-seconds/10213693

I assume that if I continue on and submit a report it won't be of any use to you because the diagnostic analyzer runner is the thing that we want to be tracing here.

**Comment by CyrusNajmabadi** [comment]
Can you just use perfview and collect a trace manually?

**Comment by reduckted** [comment]
@mavasani I managed to create a ticket: https://developercommunity.visualstudio.com/t/Error-squiggles-disappear-when-there-are/10214653

**Comment by mavasani** [comment]
> @mavasani I managed to create a ticket: https://developercommunity.visualstudio.com/t/Error-squiggles-disappear-when-there-are/10214653

Thanks @reduckted. Unfortunately, the CPU stacks for the collected trace does not show any diagnostic analysis running in the background. @CyrusNajmabadi can also check if you see anything interesting in the collected trace?

**Comment by mavasani** [comment]
Reported in https://developercommunity.visualstudio.com/t/Intellisense-squiggles-out-of-sync/10206113

**Comment by mavasani** [comment]
Reported in https://developercommunity.visualstudio.com/t/Intellisense-in-VS2022-takes-50-seconds/10213693

**Comment by CyrusNajmabadi** [comment]
Fixed with https://github.com/dotnet/roslyn/pull/65538

**Comment by mavasani** [comment]
@CyrusNajmabadi just to confirm, the fix should be present in 17.4.3 servicing release, correct?

**Comment by CyrusNajmabadi** [comment]
@arkalyanms ^

**Comment by arkalyanms** [comment]
Yes the push diagnostics went into 17.4.3.

**Comment by kirsan31** [comment]
Wow, I've just upgraded from 17.3.6 to 17.4.2 and fell into this nightmare. Even if I just change the comment (2.2k lines file), all errors instantly disappear and appear only after 30+ seconds 🤪
After a long search on developercommunity, came across this thread.
I really hope this gets fixed 17.4.**3**. 🙏
